test/jdk/java/util/zip/ZipOutputStream/CloseWrappedStream.java

```diff
@@ -0,0 +1,229 @@
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/* @test
+   @bug 8313739
+   @summary Verify that ZipOutputStream closes the wrapped stream even after failed writes
+   @run junit CloseWrappedStream
+   */
+
+import org.junit.jupiter.api.Test;
+
+import java.io.*;
+import java.nio.charset.StandardCharsets;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipOutputStream;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+public class CloseWrappedStream {
+
+    /**
+     * Verify that closing a ZipOutputStream closes the wrapped output stream,
+     * also when the wrapped stream throws while remaining data is flushed
+     */
+    @Test
+    public void exceptionDuringFinish() {
+        // A wrapped stream which should be closed even after a write failure
+        WrappedOutputStream wrappedStream = new WrappedOutputStream();
+
+        IOException exception = assertThrows(IOException.class, () -> {
+            try (ZipOutputStream zo = new ZipOutputStream(wrappedStream)) {
+                zo.putNextEntry(new ZipEntry("file.txt"));
+                zo.write("hello".getBytes(StandardCharsets.UTF_8));
+                // Make finish() throw IOException
+                wrappedStream.failOnWrite = true;
+            } // Close throws when deflated data is flushed to wrapped stream
+        });
+
+        // Check that we failed with the expected message
+        assertEquals(WrappedOutputStream.WRITE_MSG, exception.getMessage());
+
+        // Verify that the wrapped stream was closed once
+        assertEquals(1, wrappedStream.timesClosed,
+                "Expected wrapped output stream to be closed once");
+    }
+
+    /**
+     * Sanity check that the wrapped stream is closed also for the
+     * normal case where the wrapped stream does not throw
+     *
+     * @throws IOException if an unexpected IOException occurs
+     */
+    @Test
+    public void noExceptions() throws IOException {
+
+        WrappedOutputStream wrappedStream = new WrappedOutputStream();
+
+        try (ZipOutputStream zo = new ZipOutputStream(wrappedStream)) {
+            zo.putNextEntry(new ZipEntry("file.txt"));
+            zo.write("hello".getBytes(StandardCharsets.UTF_8));
+        }
+
+        // Verify that the wrapped stream was closed once
+        assertEquals(1, wrappedStream.timesClosed,
+                "Expected wrapped output stream to be closed once");
+    }
+
+    /**
+     * Check that the exception handling is correct when the
+     * wrapped stream throws while being closed
```

Comment:
This comment could use a bit of wordsmithing to indicate what "correct" means

Reply:
> This comment could use a bit of wordsmithing to indicate what "correct" means

It's hard to write good prose for these tricky error scenarios. But just saying "correct" without defining it is a bit too lazy, yes :-)

Please take a look at the last commit 39a14de where I've expanded "correct" in the test comments.